cluster-autoscaler/core/scaledown/actuation/delete_in_batch_test.go

```diff
@@ -0,0 +1,236 @@
+		},
+		{
+			name:           "Unsuccessful remove",
+			numNodes:       5,
+			failedDeletion: 1,
+			addNgToBucket:  true,
+		},
+	}
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			test := test
+			fakeClient := &fake.Clientset{}
+			fakeLogRecorder, _ := clusterstate_utils.NewStatusMapRecorder(fakeClient, "kube-system", kube_record.NewFakeRecorder(5), false, "my-cool-configmap")
+
+			failedNodeDeletion := make(map[string]bool)
+			deletedNodes := make(chan string, 10)
+			notDeletedNodes := make(chan string, 10)
+			// Hook node deletion at the level of cloud provider, to gather which nodes were deleted, and to fail the deletion for
+			// certain nodes to simulate errors.
+			provider := testprovider.NewTestCloudProvider(nil, func(nodeGroup string, node string) error {
+				if failedNodeDeletion[node] {
+					notDeletedNodes <- node
+					return fmt.Errorf("SIMULATED ERROR: won't remove node")
+				}
+				deletedNodes <- node
+				return nil
+			})
+
+			fakeClient.Fake.AddReactor("update", "nodes",
+				func(action core.Action) (bool, runtime.Object, error) {
+					update := action.(core.UpdateAction)
+					obj := update.GetObject().(*apiv1.Node)
+					return true, obj, nil
+				})
+
+			ctx, err := NewScaleTestAutoscalingContext(config.AutoscalingOptions{}, fakeClient, nil, provider, nil, nil)
+			clusterStateRegistry := clusterstate.NewClusterStateRegistry(provider, clusterstate.ClusterStateRegistryConfig{}, fakeLogRecorder, NewBackoff())
+			if err != nil {
+				t.Fatalf("Couldn't set up autoscaling context: %v", err)
+			}
+
+			ng := "ng"
+			provider.AddNodeGroup(ng, 1, 10, test.numNodes)
+
+			d := NodeDeletionBatcher{
+				ctx:                   &ctx,
+				clusterState:          clusterStateRegistry,
+				nodeDeletionTracker:   deletiontracker.NewNodeDeletionTracker(1 * time.Minute),
+				deletionsPerNodeGroup: make(map[string][]*apiv1.Node),
+				drainedNodeDeletions:  make(map[string]bool),
+			}
+			nodes := generateNodes(test.numNodes, ng)
+			failedDeletion := test.failedDeletion
+			for _, node := range nodes {
+				if failedDeletion > 0 {
+					failedNodeDeletion[node.Name] = true
+					failedDeletion -= 1
+				}
+				provider.AddNode(ng, node)
+			}
+			if test.addNgToBucket {
+				for _, node := range nodes {
+					node.Spec.Taints = append(node.Spec.Taints, apiv1.Taint{
+						Key:    deletetaint.ToBeDeletedTaint,
+						Effect: apiv1.TaintEffectNoSchedule,
+					})
+					_, _, err := d.addNodeToBucket(node, true)
+					if err != nil {
+						t.Errorf("addNodeToBucket return error %q when addidng node %v", err, node)
+					}
+				}
+			}
+
+			err = d.remove(ng)
+			if test.err {
+				if err == nil {
+					t.Errorf("remove() should return error, but return nil")
+				}
+				return
+			}
+			if err != nil {
+				t.Errorf("remove() return error, but shouldn't")
+			}
+			if test.failedDeletion == 0 {
+				gotDeletedNodes := []string{}
+				for i := 0; i < test.numNodes; i++ {
```

Comment:
nit: The current implementation only checks the deleted nodes if `test.failedDeletion != 0`, which is not ideal. I think this could all be simplified to:
```

for i := 0; i < test.numNodes; i++ {
  select {
    case deletedNode := <-deletedNodes:
	gotDeletedNodes = append(gotDeletedNodes, deletedNode)
    case notDeletedNode := <-notDeletedNodes:
	notDeletedNodes = append(notDeletedNodes, notDeletedNode)
    case <-time.After(4 * time.Second):
	t.Errorf("Timeout while waiting for deleted nodes.")
}
```

Also, why aren't we asserting `gotDeletedNodes`? It's essentialy not used at all, right?

Reply:
Because when we have failed request cloud provider stop delete process and return error. It may happen that the first node in node delete loop is failed and so all other nodes won't appear in deletedNode

Reply:
Ahh got it, we need to have the different test cloud provider implementation I mentioned in another comment. Let's do that in a follow-up.